src/fabric/annealing.rs

```diff
@@ -94,56 +172,134 @@ impl MeasurePair {
         self.pairs.extend(new_pairs);
     }
+
+    fn bow_tie_pulls(mut self) -> impl Iterator<Item=Pair> {
+        for interval in self.intervals.values() {
+            if interval.role != Role::Push {
+                continue;
+            }
+            let mut meeting_pairs: Vec<(Path, Path)> = vec![];
+            for alpha_path in self.paths_for(interval.alpha_index, 2) {
+                for omega_path in self.paths_for(interval.omega_index, 2) {
+                    if alpha_path.last() == omega_path.last() { // second one is the bridge
+                        meeting_pairs.push((alpha_path.clone(), omega_path.clone()))
+                    }
+                }
+            }
+            let [(alpha1, omega1), (alpha2, omega2)] = meeting_pairs.as_slice() else {
+                continue;
+            };
+            let diagonals = [
+                (alpha1.last_joint(), omega2.last_joint()),
+                (alpha2.last_joint(), omega1.last_joint())
+            ];
+            let candidates: Vec<(usize, usize)> = diagonals
+                .iter()
+                .filter_map(|&(a, b)| {
+                    match (self.joints[a].across_push(), self.joints[b].across_push()) {
+                        (Some(joint_a), Some(joint_b)) => {
+                            (!self.interval_exists(joint_a, joint_b)).then_some((a, b))
+                        }
+                        _ => None
+                    }
+                })
+                .collect();
+            if let &[(alpha_index, omega_index)] = candidates.as_slice() {
+                let link = Link::Pull {ideal: interval.ideal() / 3.0};
+                let pair = Pair { alpha_index, omega_index, link};
+                self.pairs.insert(pair.key(), pair);
+            }
+        }
+        self.pairs.into_values()
+    }
+
+    fn interval_exists(&self, a: usize, b: usize) -> bool {
+        if a < b {
+            self.intervals.contains_key(&(a, b))
+        } else {
+            self.intervals.contains_key(&(b, a))
+        }
+    }
+
+    fn get_interval(&self, a: usize, b: usize) -> Option<&Interval> {
+        if a < b {
+            self.intervals.get(&(a, b))
+        } else {
+            self.intervals.get(&(b, a))
+        }
+    }
+
+    fn paths_for(&self, joint_index: usize, max_size: usize) -> Vec<Path> {
+        let paths: Vec<Path> = self.joints[joint_index].pulls
+            .iter()
+            .map(|pull| Path::new(joint_index, *pull))
+            .collect();
+        self.paths_via_pulls(&paths, 1, max_size)
+    }
+
+    fn paths_via_pulls(&self, paths: &[Path], size: usize, max_size: usize) -> Vec<Path> {
+        if size == max_size {
+            paths.to_vec()
+        } else {
+            let bigger: Vec<Path> = paths
+                .iter()
+                .flat_map(|path| self.joints[path.last_joint()].extended_paths(path))
+                .collect();
+            self.paths_via_pulls(&bigger, size + 1, max_size)
+        }
```

Comment:
You could define `fn paths_via_pulls(..)` inside `paths_for(..)` if it's only ever used internally. Cleans up your impl interface

Reply:
Since it's one of those fancy recursive functions that needs an initial value..

Reply:
We just learned that internal functions are fine except that they may not be recursive.